File: Python/278. First Bad Version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
#         high, low = n, 1
#         while high > low:
#             mid = (high + low) // 2
#             if isBadVersion(mid): high = mid
#             else: low = mid + 1
#         return low

        high, low = n, 1
        while high >= low:
            mid = (high + low) // 2
            if isBadVersion(mid) and not isBadVersion(mid - 1):
                return mid
            elif isBadVersion(mid):
                high = mid - 1
            else:
                low = mid + 1
        return low

File: PHP/278. First Bad Version.php
<?php
/* The isBadVersion API is defined in the parent class VersionControl.
      public function isBadVersion($version){} */

class Solution extends VersionControl {
    /**
     * @param Integer $n
     * @return Integer
     */
    function firstBadVersion($n) {
        $l = 0;
        $r = $n;
        while($l <= $r){
            $m = round(($l + $r) / 2);
            if($this->isBadVersion($m)) $r = $m - 1;
            else $l = $m + 1;
        }
        return $l;
    }
}
